src/http/server/handlers/static_file_handler.cr

```diff
@@ -103,6 +100,39 @@ class HTTP::StaticFileHandler
     context.response.headers.add "Location", url
   end
 
+  private def add_cache_headers(response_headers : HTTP::Headers, file_path : String) : Nil
+    response_headers[HTTP_ETAG] = etag(file_path)
+
+    last_modified = modification_time(file_path)
+    response_headers[HTTP_LAST_MODIFIED] = HTTP.rfc1123_date(last_modified)
+  end
+
+  private def cache_request?(context : HTTP::Server::Context, file_path : String) : Bool
+    # According to RFC 7232:
+    # A recipient must ignore If-Modified-Since if the request contains an If-None-Match header field
+    if if_none_match = context.request.headers[HTTP_IF_NONE_MATCH]?
+      if_none_match == context.response.headers[HTTP_ETAG]
+    elsif if_modified_since = context.request.headers[HTTP_IF_MODIFIED_SINCE]?
+      header_time = HTTP.parse_time(if_modified_since)
+      # File mtime probably has a higher resolution than the header value.
+      # An exact comparison might be slightly off, so we add 1s padding.
+      # Static files should generally not be modified in subsecond intervals, so this is perfectly safe.
+      # This might be replaced by a more sophisticated time comparison when it becomes available.
+      last_modified = modification_time(file_path)
+      !!(header_time && last_modified <= header_time + 1.second)
+    else
+      false
+    end
+  end
+
+  private def etag(file_path)
```

Comment:
This should take a `modification_time` so that we only call `File.info(path).modification_time` once, this saves performance and avoids race conditions where the etag and last modified are different.

Reply:
Race condition doesn't matter, but performance and reusability 👍

Reply:
Could even move this as a property of `Response` for even more reusability.